drv/stm32xx-gpio-common/src/lib.rs

```diff
@@ -0,0 +1,174 @@
+// This Source Code Form is subject to the terms of the Mozilla Public
+// License, v. 2.0. If a copy of the MPL was not distributed with this
+// file, You can obtain one at https://mozilla.org/MPL/2.0/.
+
+//! GPIO-related things needed by all STM32 parts.
+
+#![no_std]
+
+use userlib::FromPrimitive;
+use zerocopy::AsBytes;
+
+#[cfg(feature = "server-support")]
+pub mod server;
+
+/// Enumerates the GPIO ports available on this chip, from the perspective of
+/// driver software. This does not mean the GPIO port is physically available on
+/// pins of the package -- we don't model package differences.
+#[derive(Copy, Clone, Debug, Eq, PartialEq, FromPrimitive, AsBytes)]
+#[repr(u8)]
+pub enum Port {
+    A = 0,
+    B,
+    C,
+    D,
+    #[cfg(feature = "has-port-gpioe")]
+    E,
+    #[cfg(feature = "has-port-gpiof")]
+    F,
+    #[cfg(feature = "has-port-gpiog")]
+    G,
+    #[cfg(feature = "has-port-gpioh")]
+    H,
+    #[cfg(feature = "has-port-gpioi")]
+    I,
+    #[cfg(feature = "has-port-gpioj")]
+    J,
+    #[cfg(feature = "has-port-gpiok")]
+    K,
+}
+
+impl Port {
+    /// Turns a `Port` into a `PinSet` containing one pin, number `index`.
+    pub const fn pin(self, index: usize) -> PinSet {
```

Comment:
Does `#[inline(always)]` improve code size here?

Reply:
I've never seen this routine _not_ get inlined. In practice that's probably because it's almost always called in `const` context though. I can add a directive.